flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/StateRequestsBuffer.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * A buffer to hold state requests to execute state requests in batch.
+ *
+ * @param <R> the type of the record
+ * @param <K> the type of the key
+ */
+@NotThreadSafe
+public class StateRequestsBuffer<R, K> {
+    /**
+     * The state requests in this buffer could be executed when the buffer is full or configured
+     * batch size is reached. All operations on this buffer must be invoked in task thread.
+     */
+    final Deque<StateRequest<K, ?, ?>> activeQueue;
+
+    /**
+     * The requests in that should wait until all preceding records with identical key finishing its
+     * execution. After which the queueing requests will move into the active buffer. All operations
+     * on this buffer must be invoked in task thread.
+     */
+    final Map<K, Deque<StateRequest<K, ?, ?>>> blockingQueue;
+
+    /** The number of state requests in blocking queue. */
+    int blockingQueueSize;
+
+    public StateRequestsBuffer() {
+        this.activeQueue = new LinkedList<>();
+        this.blockingQueue = new HashMap<>();
+        this.blockingQueueSize = 0;
+    }
+
+    void enqueueToActive(StateRequest<K, ?, ?> request) {
+        activeQueue.add(request);
+    }
+
+    void enqueueToBlocking(StateRequest<K, ?, ?> request) {
+        blockingQueue
+                .computeIfAbsent(request.getRecordContext().getKey(), k -> new LinkedList<>())
+                .add(request);
+        blockingQueueSize++;
+    }
+
+    /**
+     * Try to migrate one state request with specific key from blocking queue to active queue.
+     *
+     * @param key The key to release, the other records with this key is no longer blocking.
+     * @return The first record context with the same key in blocking queue, null if no such record.
+     */
+    @Nullable
+    RecordContext<R, K> tryMigrateSpecificOne(K key) {
+        if (!blockingQueue.containsKey(key)) {
+            return null;
+        }
+
+        StateRequest<K, ?, ?> stateRequest = blockingQueue.get(key).getFirst();
+        activeQueue.add(stateRequest);
+        blockingQueue.get(key).removeFirst();
+        if (blockingQueue.get(key).isEmpty()) {
+            blockingQueue.remove(key);
+        }
+        blockingQueueSize--;
+        return (RecordContext<R, K>) stateRequest.getRecordContext();
+    }
+
+    /**
+     * Get the number of state requests of blocking queue in constant-time.
+     *
+     * @return the number of state requests of blocking queue.
+     */
+    @VisibleForTesting
+    int blockingQueueSize() {
+        return blockingQueueSize;
+    }
+
+    /**
+     * Get the number of state requests of active queue in constant-time.
+     *
+     * @return the number of state requests of active queue.
+     */
+    @VisibleForTesting
+    int activeQueueSize() {
+        return activeQueue.size();
+    }
+
+    /** Clear all state requests in the active queue. */
+    void clearActiveQueue() {
+        activeQueue.clear();
+    }
+
+    List<StateRequest<?, ?, ?>> nextToRun() {
```

Comment:
I'd suggest we combine `clearActiveQueue` and `nextToRun` together and provide a new method that could `pop` a given number of `StateRequest`s to `AEC`.